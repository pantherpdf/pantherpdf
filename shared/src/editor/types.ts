/**
 * types.ts
 * types used in report editor
 */


import { IconDefinition } from '@fortawesome/fontawesome-svg-core'
import type { FunctionComponent, ReactNode } from 'react'
import type { FormulaHelper } from './compile'
import { TReport, TReportShort, TData, TTransformData, ApiEndpoints, TDataCompiled } from '../types'



export type TName = string | {[key: string]: string}
export interface TFontAwesomeIcon { fontawesome: IconDefinition }




// to help construct tests
// to force specific children type
export type ForceChildren<T> = T | {[key:string]: any, children: ForceChildren<T>[]}






export type ReportForceChildren<T> = TReport & { children: ForceChildren<T>[] }





export type TDragObj = {type:'wid', wid: number[]} | {type:'widget', widget: TData} | {type:'widgets', widgets: TData[]}

export interface GeneralProps {
	getOriginalSourceData: (report: TReport) => any,
	overrideSourceData: (data: any) => void,
	isOverridenSourceData: boolean,
	sourceData: any,
	sourceErrorMsg?: string | undefined,
	refreshSourceData: (report: TReport) => void,
	api: ApiEndpoints,

	report: TReport,
	setReport: (report: TReport) => Promise<void>,
	deleteReport: () => void,

	selected: number[] | null,
	setSelected: React.Dispatch<React.SetStateAction<number[]|null>>,

	renderWidget: (child: TData, parents: number[]) => ReactNode,
	renderWidgets: (children: TData[], parents: number[]) => ReactNode,
	dragWidgetStart: (e: React.DragEvent<HTMLDivElement>, dragObj: TDragObj) => void,
	dragWidgetEnd: (e: React.DragEvent<HTMLDivElement>) => void,
	drop: (e: React.DragEvent<HTMLDivElement>, dest: number[]) => void,
}

export interface ItemRendeProps extends GeneralProps {
	item: TData,
	setItem: (itm: TData) => void,
	wid: number[],
}

export interface ItemRendeFinalHelper {
	renderItem: (item: TDataCompiled, helper: ItemRendeFinalHelper) => ReactNode,
	renderChildren: (chs: TDataCompiled[], helper: ItemRendeFinalHelper) => ReactNode,
}

export interface ItemRendeFinalProps extends ItemRendeFinalHelper {
	item: TDataCompiled,
}

export interface CompileHelper {
	formulaHelper: FormulaHelper,
	evalFormula: (txt: string) => Promise<any>,
	compileChildren: (arr1: TData[], helper: CompileHelper) => Promise<TDataCompiled[]>,
	wid: number[],
	report: TReport,
	api?: ApiEndpoints,
}

export interface NewItemProps {
	report: TReport,
}

export interface Widget {
	name: TName,
	icon: TFontAwesomeIcon,
	newItem: (props: NewItemProps) => Promise<TData>,
	compile: (dt: any, helper: CompileHelper) => Promise<TDataCompiled>,
	RenderProperties?: FunctionComponent<ItemRendeProps>,
	Render: FunctionComponent<ItemRendeProps>,
	RenderFinal: FunctionComponent<ItemRendeFinalProps>,
	canAdd?: boolean,
	canDrag?: boolean,
	canSelect?: boolean,
}


export interface TransformRendeProps extends GeneralProps {
	item: TTransformData,
	setItem: (itm: TTransformData) => void,
	index: number,
}


export interface TTransformWidget {
	name: TName,
	newItem: () => Promise<TTransformData>,
	transform: (dt: any, item: TTransformData) => Promise<any>,
	Editor: FunctionComponent<TransformRendeProps>,
}

