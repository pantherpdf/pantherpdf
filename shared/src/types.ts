//import type { ObjectId } from "mongodb"

// helper for converting tuple into type
type Narrowable = string | number | boolean | symbol | object | {} | void | null | undefined;
const tuple = <T extends Narrowable[]>(...args: T)=>args;


export type WithId<T> = T & {_id: string}

export interface IUser {
	email: string,
	name: string,
	anonymousId?: string,
}
export type IUserWithId = WithId<IUser>

export interface ISession {
	email: string,
	sid: string,
}

export interface IReportGenerated {
	email: string,
	time: string,
	html: string,
	reportId: string,
	accessKey: string,
}

export interface TData {
	[key: string]: any,
	type: string,
	children: TData[],
}

export const TargetOptions = tuple('pdf', 'json', 'csv-excel-utf-8', 'csv-windows-1250');
export type TargetOption = (typeof TargetOptions)[number];
export function TargetOptionTypeGuard(r: any): r is TargetOption {
	if (typeof r !== 'string')
		return false
	if ((TargetOptions as string[]).indexOf(r) === -1)
		return false
	return true
}

export interface TTransformData {
	[key: string]: any,
	type: string,
	comment: string,
}

export interface TReport {
	_id: string,
	name: string,
	email: string,
	time: string,
	target: TargetOption,
	children: TData[],
	transforms: TTransformData[],
	properties: {
		font?: any,
		margin?: [number, number, number, number],
		fileName?: string,
		paperWidth?: number,
		paperHeight?: number,
		lang?: string,
	}
}
export type TReportWithoutId = Omit<TReport, '_id'>

export interface TDataCompiled {
	[key: string]: any,
	type: string,
	children: TDataCompiled[],
}

export interface TReportCompiled extends Omit<TReport, 'children'> {
	children: TDataCompiled[],
}

export function TDataTypeGuard(r: any): r is TData {
	if (typeof r.type !== 'string' || r.type.length == 0)
		return false
	if (!Array.isArray(r.children))
		return false
	for (const ch of r.children) {
		if (!TDataTypeGuard(ch))
			return false
	}
	return true
}

export function ReportTypeGuard(r: any): r is TReport {
	if (typeof r != 'object')
		return false
	if (typeof r.email !== 'string')
		return false
	if (typeof r.name !== 'string')
		return false
	if (r.target !== 'pdf' && r.target !== 'json' && r.target !== 'csv-excel-utf-8' && r.target !== 'csv-windows-1250')
		return false
	if (!Array.isArray(r.children))
		return false
	for (const ch of r.children) {
		if (!TDataTypeGuard(ch))
			return false
	}
	if (typeof r.properties !== 'object')
		return false
	return true
}

export interface TReportShort {
	_id: string,
	name: string,
	target: TargetOption,
}




export interface TFile {
	email: string,
	name: string,
	mimeType: string,
	uploadTime: string,
	modifiedTime: string,
	data: Buffer,
	size: number,
}

export interface TFileShort {
	name: string,
	mimeType: string,
	uploadTime: string,
	modifiedTime: string,
	size: number,
}


export interface ApiEndpoints {
	reportGet: (id: string) =>  Promise<ReportResponse2>,
	files: () =>  Promise<FilesResponse2>,
	filesDelete: (name: string) =>  Promise<void>,
	filesUpload: (file: File, data: FileUploadData, cbProgress: (prc: number) => void) =>  Promise<void>,
	filesDownloadUrl: (name: string) => string,
}





















export interface ErrorResponse { msg: string }


// loginAnonymous
export interface LoginAnonymousRequest { anonymousId?: string }
export function LoginAnonymousRequestTypeGuard(r: any): r is LoginAnonymousRequest {
	if (typeof r != 'object')
		return false
	if ('anonymousId' in r) {
		if (typeof r.anonymousId != 'string')
			return false
		if (r.anonymousId.length == 0)
			return false
	}
	return true
}
export type LoginAnonymousResponse = { sid: string, anonymousId: string } | ErrorResponse


// userData
export type UserDataResponse = { user: IUser, reports: TReportShort[] } | ErrorResponse


// keys
export interface IKey {
	key: string,
	name: string,
	email: string,
	time: string,
}
export interface IKeyPublicShort {
	name: string,
	time: string,
}
export type KeysResponse = IKeyPublicShort[] | ErrorResponse

export type KeyAddRequest = { name: string }
export function KeyAddRequestTypeGuard(r: any): r is KeyAddRequest {
	if (typeof r != 'object')
		return false
	if (!('name' in r) || typeof r.name != 'string')
		return false
	return true
}
export type KeyAddResponse = (IKeyPublicShort&{ key: string, }) | ErrorResponse

export type KeyRemoveRequest = { name: string }
export function KeyRemoveRequestTypeGuard(r: any): r is KeyRemoveRequest {
	if (typeof r != 'object')
		return false
	if (!('name' in r) || typeof r.name != 'string')
		return false
	return true
}
export type KeyRemoveResponse = { } | ErrorResponse

export type KeyUpdateRequest = { nameOld: string, nameNew: string }
export function KeyUpdateRequestTypeGuard(r: any): r is KeyUpdateRequest {
	if (typeof r != 'object')
		return false
	if (!('nameOld' in r) || typeof r.nameOld != 'string')
		return false
	if (!('nameNew' in r) || typeof r.nameNew != 'string')
		return false
	return true
}


// events
export interface IEvent {
	email?: string,
	sid?: string,
	key?: string,
	ip?: string,
	userAgent?: string,
	time: string,
	type: string,
	[x: string]: any 
}


// Reports
export interface ReportNewRequest {
	name: string,
}
export function ReportNewRequestTypeGuard(r: any): r is ReportNewRequest {
	if (typeof r != 'object')
		return false
	if (!('name' in r) || typeof r.name != 'string')
		return false
	return true
}
export type ReportNewResponse = TReportShort | ErrorResponse

export interface ReportRemoveRequest {
	id: string,
}
export function ReportRemoveRequestTypeGuard(r: any): r is ReportRemoveRequest {
	if (typeof r != 'object')
		return false
	if (!('id' in r) || typeof r.id != 'string')
		return false
	return true
}
export type ReportRemoveResponse = { } | ErrorResponse

export type ReportResponse2 = { obj: WithId<TReport> }
export type ReportResponse = ReportResponse2 | ErrorResponse






export type FilesResponse2 = { files: TFileShort[] }
export type FilesResponse = FilesResponse2 | ErrorResponse

export interface FileUploadData {
	name: string,
	modifiedTime: string,
	mimeType: string,
}

export function FileUploadDataTypeGuard(r: any): r is FileUploadData {
	if (typeof r !== 'object')
		return false
	if (!('name' in r) || typeof r.name !== 'string')
		return false
	if (!('modifiedTime' in r) || typeof r.modifiedTime !== 'string')
		return false
	const rgx = /^[\d]{4}-[\d]{2}-[\d]{2}T[\d]{2}:[\d]{2}:[\d]{2}Z$/gm
	if (!r.modifiedTime.match(rgx))
		return false
	if (!('mimeType' in r) || typeof r.mimeType !== 'string')
		return false
	return true
}

export type FileUploadResponse = {file: TFileShort} | ErrorResponse
