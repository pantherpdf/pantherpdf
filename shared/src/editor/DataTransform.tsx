/**
 * DataTransform.tsx
 * Manage data transformations. Add, delete, edit TTransformWidget.
 */


import React, { useState } from 'react'
import { TReport, TTransformData } from '../types'
import { GeneralProps } from './types'
import { Dropdown, Button, ButtonGroup, Modal } from 'react-bootstrap'
import Trans, { TransName } from '../translation'
import getTransform, { allTransforms } from '../transforms/allTransforms'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faArrowDown, faArrowUp, faEdit, faPlus, faTimes, faTrash, faRetweet } from '@fortawesome/free-solid-svg-icons'
import ObjectExplorer from './ObjectExplorer'
import FileSelect from '../FileSelect'


/**
 * Function for calling all transformations
 * @func
 * @param {inputData} any - Input data
 * @param {len} number - Number of transformations to apply
 */
export async function transformData(inputData: any, report: TReport, len?: number) {
	if (len === undefined) {
		len = report.transforms.length
	}
	for (let i=0; i<len; ++i) {
		const w = report.transforms[i]
		const comp = getTransform(w.type)
		inputData = await comp.transform(inputData, w)
	}
	return inputData
}


interface TransformItemProps extends GeneralProps {
	item: TTransformData,
	index: number,
	showData: (len: number) => void,
	up: (idx: number) => void,
	down: (idx: number) => void,
	itemDelete: (idx: number) => void,
	openEditor: (idx: number) => void,
}
function TransformItem(props: TransformItemProps) {
	const comp = getTransform(props.item.type)
	return <Dropdown
		as={ButtonGroup}
		className='d-flex'
		size='sm'
	>
		<button className='btn btn-sm btn-outline-secondary flex-fill text-left' onClick={()=>props.showData(props.index+1)}>
			{TransName(comp.name)} <small className='text-truncate d-inline-block ml-3 text-muted' style={{maxWidth:'5rem', position:'relative', top:'0.25rem'}}>{props.item.comment}</small>
		</button>
		<Dropdown.Toggle style={{maxWidth:'3rem'}} split variant="outline-secondary" id="dropdown-custom-2" />
		<Dropdown.Menu>
			<Dropdown.Item onClick={()=>props.openEditor(props.index)}>
				<FontAwesomeIcon icon={faEdit} className='me-2' fixedWidth />
				{Trans('edit')}
			</Dropdown.Item>
			<Dropdown.Item onClick={e=>{e.preventDefault(); props.up(props.index)}} disabled={props.index===0}>
				<FontAwesomeIcon icon={faArrowUp} className='me-2' fixedWidth />
				{Trans('up')}
			</Dropdown.Item>
			<Dropdown.Item onClick={e=>{e.preventDefault(); props.down(props.index)}} disabled={props.index+1===props.report.transforms.length}>
				<FontAwesomeIcon icon={faArrowDown} className='me-2' fixedWidth />
				{Trans('down')}
			</Dropdown.Item>
			<Dropdown.Divider />
			<Dropdown.Item onClick={e=>{e.preventDefault(); props.itemDelete(props.index)}}>
				<FontAwesomeIcon icon={faTrash} className='me-2' fixedWidth />
				{Trans('delete')}
			</Dropdown.Item>
		</Dropdown.Menu>
	</Dropdown>
}



interface TEdit {
	index: number,
	data: TTransformData,
}

interface TSourceDataShow {
	length?: number,
	data: any,
	errorMsg?: string,
}

export default function DataTransform(props: GeneralProps) {
	const [shownModalInsert, setShownModalInsert] = useState<boolean>(false)
	const [showEdit, setShowEdit] = useState<TEdit | null>(null)
	const [showData, setShowData] = useState<TSourceDataShow | null>(null)
	const [shownModalUpload, setShownModalUpload] = useState<boolean>(false)
	const [uploadUrl, setUploadUrl] = useState<string>('')

	async function doShowData(len: number) {
		let dt
		try {
			dt = props.getOriginalSourceData();
		}
		catch(e) {
			const msg = e instanceof Error ? e.message : 'Unknown error'
			setShowData({data: null, errorMsg: msg})
			return
		}
		
		let dt2
		try {
			dt2 = await transformData(dt, props.report, len)
		}
		catch(e) {
			const msg = e instanceof Error ? e.message : 'Unknown error'
			setShowData({data: null, errorMsg: msg})
			return
		}
		setShowData({data: dt2, length: len})
	}

	async function itemAdd(key: string) {
		const cmp = getTransform(key)
		const dt = await cmp.newItem()
		const report2: TReport = {...props.report, transforms:[...props.report.transforms, dt]}
		await props.setReport(report2)
		setShownModalInsert(false)
		setShowEdit({
			index: report2.transforms.length-1,
			data: JSON.parse(JSON.stringify(dt))
		})
	}

	async function itemUp(idx: number) {
		const report2: TReport = {...props.report, transforms:[...props.report.transforms]}
		const tmp = report2.transforms[idx-1]
		report2.transforms[idx-1] = report2.transforms[idx]
		report2.transforms[idx] = tmp
		return props.setReport(report2)
	}

	async function itemDown(idx: number) {
		const report2: TReport = {...props.report, transforms:[...props.report.transforms]}
		const tmp = report2.transforms[idx+1]
		report2.transforms[idx+1] = report2.transforms[idx]
		report2.transforms[idx] = tmp
		return props.setReport(report2)
	}

	async function itemDelete(idx: number) {
		const report2: TReport = {...props.report, transforms:[...props.report.transforms]}
		report2.transforms.splice(idx, 1)
		return props.setReport(report2)
	}

	function itemOpenEditor(idx: number) {
		setShowEdit({
			index: idx,
			data: JSON.parse(JSON.stringify(props.report.transforms[idx]))
		})
	}


	async function editSave() {
		if (!showEdit)
			return
		const report2: TReport = {...props.report, transforms:[...props.report.transforms]}
		report2.transforms[showEdit.index] = showEdit.data
		await props.setReport(report2)
		setShowEdit(null)
		props.refreshSourceData(report2)
	}

	function editChange(data: TTransformData) {
		if (showEdit)
			setShowEdit({...showEdit, data})
	}

	const editCmp = showEdit ? getTransform(showEdit.data.type) : null

	return <>
		<Dropdown
			as={ButtonGroup}
			className='d-flex'
			size='sm'
		>
			<button className='btn btn-sm btn-outline-secondary flex-fill text-left' onClick={()=>doShowData(0)}>
				{Trans('source data')}
			</button>
			<Dropdown.Toggle style={{maxWidth:'3rem'}} split variant="outline-secondary" id="dropdown-custom-2" />
			<Dropdown.Menu>
				<Dropdown.Item onClick={() => {
					setShownModalUpload(true)
				}}>
					<FontAwesomeIcon icon={faRetweet} className='me-2' fixedWidth />
					{Trans('replace')}
				</Dropdown.Item>
				{props.indexOverridenSourceData>0 && <Dropdown.Item onClick={() => {
					props.overrideSourceData(null)
				}}>
					<FontAwesomeIcon icon={faTimes} className='me-2' fixedWidth />
					{Trans('reset')}
				</Dropdown.Item>}
			</Dropdown.Menu>
		</Dropdown>


		{props.report.transforms.map((item, idx) => <TransformItem
			{...props}
			item={item}
			index={idx}
			key={idx}
			showData={doShowData}
			up={itemUp}
			down={itemDown}
			itemDelete={itemDelete}
			openEditor={itemOpenEditor}
		/>)}
		<div className=''>
			<button className='btn btn-sm btn-outline-secondary mt-2' onClick={()=>setShownModalInsert(true)}>
				<FontAwesomeIcon icon={faPlus} className='me-2' />
				{Trans('transform')}
			</button>
		</div>

		{/* UPLOAD JSON */}
		<Modal show={shownModalUpload} onHide={() => setShownModalUpload(false)}>
			<Modal.Header closeButton>
				<Modal.Title>
					Upload JSON data
				</Modal.Title>
			</Modal.Header>
			<Modal.Body>
				<div className="mb-3">
					<label htmlFor="uploadUrl" className="form-label">
						Url
					</label>
					<input
						type="text"
						placeholder='https://www.example.com/file.json'
						className="form-control"
						id="uploadUrl"
						value={uploadUrl}
						onChange={e => setUploadUrl(e.currentTarget.value)}
					/>
				</div>
				
				<hr />

				<FileSelect
					onSelect={async fls => {
						if (fls.length === 0) {
							return
						}
						const f = fls[0]
						try {
							const reader = new FileReader()
							reader.onload = (e2) => {
								if (!e2.target || typeof e2.target.result !== 'string') {
									return
								}
								const dt = JSON.parse(e2.target.result)
								props.overrideSourceData(dt)
							}
							reader.readAsText(f)
						}
						catch(e) {
							alert(`Error: ${String(e)}`)
							return
						}
						setShownModalUpload(false)
					}}
				/>
			</Modal.Body>
			<Modal.Footer>
				<Button
					variant="secondary"
					onClick={() => {
						setShownModalUpload(false)
					}}
				>
					{Trans('cancel')}
				</Button>
				<Button
					variant="primary"
					onClick={async () => {
						let js: any
						try {
							const r = await fetch(uploadUrl)
							if (!r.ok) {
								alert('Bad response from url')
								return
							}
							js = await r.json()
						}
						catch(e) {
							alert(`Error: ${String(e)}`)
							return
						}
						setShownModalUpload(false)
						props.overrideSourceData(js)
					}}
					disabled={uploadUrl.length === 0}
				>
					Upload
				</Button>
			</Modal.Footer>
		</Modal>

		{/* ADD NEW */}
		<Modal show={shownModalInsert} onHide={() => setShownModalInsert(false)}>
			<Modal.Header closeButton>
				<Modal.Title>
					{Trans('insert data transform')}
				</Modal.Title>
			</Modal.Header>
			<Modal.Body>
				<div className="list-group">
					{Object.keys(allTransforms).map(key => {
						const w = allTransforms[key]
						return <button
							className='btn list-group-item list-group-item-action'
							key={key}
							onClick={()=>itemAdd(key)}
						>
							{TransName(w.name)}
						</button>
					})}
				</div>
			</Modal.Body>
		</Modal>

		{/* EDIT */}
		<Modal show={!!showEdit} onHide={() => setShowEdit(null)}>
			{editCmp && showEdit && <>
			<Modal.Header closeButton>
				<Modal.Title>
					{Trans('edit')} {TransName(editCmp.name)}
				</Modal.Title>
			</Modal.Header>
			<Modal.Body>
				<editCmp.Editor
					{...props}
					index={showEdit.index}
					item={showEdit.data}
					setItem={editChange}
				/>
			</Modal.Body>
			<Modal.Footer>
				<input
					type='text'
					className='form-control flex-fill w-auto'
					id='trans-edit-comment'
					value={showEdit.data.comment}
					placeholder={Trans('comment')}
					onChange={e => editChange({...showEdit.data, comment: e.currentTarget.value})}
				/>
				<Button variant="secondary" onClick={() => setShowEdit(null)}>
					{Trans('cancel')}
				</Button>
				<Button variant="primary" onClick={editSave}>
					{Trans('save')}
				</Button>
			</Modal.Footer>
			</>}
		</Modal>

		{/* DATA */}
		<Modal show={!!showData} onHide={() => setShowData(null)}>
			<Modal.Header closeButton>
				<Modal.Title>
					{Trans('data')}
				</Modal.Title>
			</Modal.Header>
			<Modal.Body>
				{showData?.data && <ObjectExplorer data={showData.data} />}
				{showData?.errorMsg && <div>{showData?.errorMsg}</div>}
			</Modal.Body>
		</Modal>
	</>
}
