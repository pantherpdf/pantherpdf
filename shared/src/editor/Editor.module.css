.boxMain {
	position: absolute;
	left: 20%;
	top: 0; /* keep 0, to prevent scroll bar from appearing when content is smaller than window */
	right: 20%;
	min-height: 100vh;
	padding: 4rem 1rem 0 1rem;
}
.boxMain:focus {
	outline: none;
}
.box1, .box2, .box3, .box4 {
	position: fixed;
	overflow-y: auto;
}
.box1 {
	left: 0;
	top: 4rem;
	bottom: calc(50% + 10px);
	width: 20%;
	border-bottom: 2px solid #999;
	border-right: 1px solid #ddd;
	padding: 0 0.5rem 10px 0.5rem;
}
.box2 {
	left: 0;
	top: 50%;
	bottom: 0;
	width: 20%;
	border-right: 1px solid #ddd;
	padding: 0 0.5rem;
}
.box3 {
	right: 0;
	top: 4rem;
	bottom: 0;
	width: 20%;
	border-left: 1px solid #ddd;
	padding: 0 0.5rem;
}


@media screen and (max-width: 560px) {
	.box1, .box2, .box3 {
		display: none;
	}
	.boxMain {
		left: 0;
		right: 0;
	}
}


.spacer {
	height: 0.5rem;
	width: 100%;
}
.dragging {
	background-color: red;
}

.widgetBox {
	border: 1px dotted #ccc;
}
.hover {
	border-color: #ff0000;
}
.selected {
	border-color: #ff0000;
	border-style: solid;
}

.nameContainer {
	background-color: rgba(0,0,0,0.03);
	box-shadow: 0px 4px 5px 0px rgba(0,0,0,0.1);
	padding: 0.5rem 0.7rem;
	margin-bottom: 1rem;
}
