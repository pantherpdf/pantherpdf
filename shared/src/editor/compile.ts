/**
 * compile.ts
 * Compile sourceData and report - evaluate all formulas
 */

import { TReport, TData } from '../types'
import type { TReportCompiled, TDataCompiled } from './types'
import FormulaEvaluate from '../formula/formula'
import getWidget from '../widgets/allWidgets'

type TOvrr = [string, any]

export class Helper {
	overrides: TOvrr[]

	constructor() {
		this.overrides = []
	}

	push(name: string, valueOrCb: any) {
		this.overrides.push([name, valueOrCb])
	}

	pop() {
		this.overrides.pop()
	}

	async getVar(name: string): Promise<any> {
		for (let i=this.overrides.length-1; i>=0; --i) {
			if (this.overrides[i][0] == name) {
				let val = this.overrides[i][1]
				if (typeof val == 'function') {
					val = await val()
				}
				return val
			}
		}
		return undefined
	}

	async evalFormula(txt: string): Promise<any> {
		return FormulaEvaluate(txt, {getVar: this.getVar.bind(this)})
	}

	async compileChildren(arr1: TData[]): Promise<TDataCompiled[]> {
		// dont use promise.all() because order of execution matters and some async operation could change it
		const arr2 = []
		for (const ch of arr1) {
			const dt = await getWidget(ch.type).compile(ch, this)
			arr2.push(dt)
		}
		return arr2
	}
}


export default async function compile(dt: TReport, obj: any): Promise<TReportCompiled> {
	const helper = new Helper()
	helper.push('obj', obj)
	const dt2: TReportCompiled = {...dt}
	dt2.children = await helper.compileChildren(dt.children)
	helper.pop()	
	return dt2
}
