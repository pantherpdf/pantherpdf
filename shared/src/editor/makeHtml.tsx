/**
 * makeHtml.ts
 * Compiled report -> html
 */


import React from 'react'
import type { TReportCompiled, ItemRendeFinalProps, TDataCompiled, ItemRendeFinalHelper } from './types'
import getWidget from '../widgets/allWidgets'


export default function makeHtml(report: TReportCompiled) {
	const helper: ItemRendeFinalHelper = {
		renderItem: (item, helper) => {
			const w = getWidget(item.type)
			return <w.RenderFinal {...helper} item={item} />
		},
		renderChildren: (chs, helper) => {
			return <div>{chs.map((item, idx) => {
				const w = getWidget(item.type)
				return <w.RenderFinal {...helper} item={item} key={idx} />
			})}</div>  // todo key
		},
	}

	const arr = helper.renderChildren(report.children, helper)
	return arr
}
