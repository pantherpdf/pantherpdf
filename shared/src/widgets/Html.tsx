/**
 * Html.tsx
 * Render html
 */


import React from 'react'
import { TData, TDataCompiled } from '../types'
import type { Widget } from '../editor/types'
import { faAlignLeft } from '@fortawesome/free-solid-svg-icons'
import BoxName from './BoxName'
import InputApplyOnEnter from './InputApplyOnEnter'


export interface HtmlData extends TData {
	type: 'Html',
	formula: string,
}

export interface HtmlCompiled extends TDataCompiled {
	type: 'Html',
	data: string,
}


export const Html: Widget = {
	name: {en: 'Html'},
	icon: {fontawesome: faAlignLeft},

	newItem: async (): Promise<HtmlData> => {
		return {
			type: 'Html',
			children: [],
			formula: '',
		}
	},

	compile: async (dt: HtmlData, helpers): Promise<HtmlCompiled> => {
		const str2 = await helpers.evalFormula(dt.formula)
		const str = (str2 !== undefined && str2 !== null && str2 !== false) ? String(str2) : ''
		return {
			type: dt.type,
			children: [],
			data: str,
		}
	},

	Render: function(props) {
		const item = props.item as HtmlData
		return <BoxName name={Html.name}>
			<div className='font-monospace'>
				{item.formula}
			</div>
		</BoxName>
	},

	RenderFinal: function(props) {
		const item = props.item as HtmlCompiled
		return <div
			dangerouslySetInnerHTML={{__html: item.data}}
		/>
	},

	RenderProperties: function(props) {
		const item = props.item as HtmlData
		return <>
			<InputApplyOnEnter
				value={item.formula}
				onChange={val=>props.setItem({...item, formula: val})}
			/>
		</>
	},
}
