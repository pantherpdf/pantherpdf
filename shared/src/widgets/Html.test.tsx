/**
 * @jest-environment node
 */

import { Html, HtmlData, HtmlCompiled } from './Html'
import compile, { Helper } from '../editor/compile'
import makeHtml from '../editor/makeHtml'
import renderer from 'react-test-renderer'
import { sampleReport } from '../editor/sampleReport'
import { ReportForceChildren } from '../editor/types'


test('Html compile', async () => {
	const dt: HtmlData = {type:'Html', formula:'data.desc', children:[]}

	const obj = { desc: '<p>Hello</p>' }
	const helper = new Helper()
	helper.push('data', obj)

	const p2 = await Html.compile(dt, helper)
	expect(p2).toBeTruthy()
	expect(p2.type).toBe('Html')
	const p = p2 as HtmlCompiled
	expect(p.data).toBe('<p>Hello</p>')
})


test('html render', async () => {
	const report: ReportForceChildren<HtmlData> = {
		...sampleReport,
		children: [
			{type:'Html', formula:'data.desc', children:[]},
		]
	}

	const data = { desc: '<p>Hello</p>' }
	const compiled = await compile(report, data)
	const html = makeHtml(compiled)
	const component = renderer.create(<>{html}</>)
	const tree = component.toJSON()
  	expect(tree).toMatchSnapshot();
})


test('Html formula==null should be empty string', async () => {
	const dt: HtmlData = {type:'Html', formula:'data.dt', children:[]}
	const helper = new Helper()
	helper.push('data', { dt: null })
	const p2 = await Html.compile(dt, helper)
	expect(p2.data).toBe('')
})

test('Html formula==false should be empty string', async () => {
	const dt: HtmlData = {type:'Html', formula:'data.dt', children:[]}
	const helper = new Helper()
	helper.push('data', { dt: false })
	const p2 = await Html.compile(dt, helper)
	expect(p2.data).toBe('')
})

test('Html formula==undefined should be empty string', async () => {
	const dt: HtmlData = {type:'Html', formula:'data.dt', children:[]}
	const helper = new Helper()
	helper.push('data', { dt: undefined })
	const p2 = await Html.compile(dt, helper)
	expect(p2.data).toBe('')
})

test('Html formula==0 should be "0"', async () => {
	const dt: HtmlData = {type:'Html', formula:'data.dt', children:[]}
	const helper = new Helper()
	helper.push('data', { dt: 0 })
	const p2 = await Html.compile(dt, helper)
	expect(p2.data).toBe('0')
})
 