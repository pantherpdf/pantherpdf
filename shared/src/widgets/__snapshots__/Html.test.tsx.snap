// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html render 1`] = `
<div>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<p>Hello</p>",
      }
    }
  />
</div>
`;
